src/tests/class_tests/openms/source/FragmentIndex_test.cpp

```diff
@@ -0,0 +1,157 @@
+}
+
+TEST_TRUE(testbuilda)
+END_SECTION
+
+
+////TEST Different Charges of the query Spectrum ////
+START_SECTION(void querySpectrum(const MSSpectrum& spectrum, SpectrumMatchesTopN& sms))
+
+
+
+std::vector<FASTAFile::FASTAEntry> entries{{"test1", "test1","MSDEREVAEAATGEDASSPPPKTEAASDPQHPAASEGAAAAAASPPLLRCLVLTGFGGYDKVKLQSRPAAPPAPGPGQLTLRLRACGLNFADLMARQGLYDRLPPLPVTPGMEGAGVVIAVGEGVSDRKAGDRVMVLNRSGMWQEEVTVPSVQTFLIPEAMTFEEAAALLVNYITAYMVLFDFGNLQPGHSVLVHMAAGGVGMAAVQLCRTVENVTVFGTASASKHEALKENGVTHPIDYHTTDYVDEIKKISPKGVDIVMDPLGGSDTAKGYNLLKPMGKVVTYGMANLLTGPKRNLMALARTWWNQFSVTALQLLQANRAVCGFHLGYLDGEVELVSGVVARLLALYNQGHIKPHIDSVWPFEKVADAMKQMQEKKNVGKVLLVPGPEKEN"}};
+AASequence protein = AASequence::fromString(entries[0].sequence);
+
+FragmentIndex queryTest;
+queryTest.build(entries);
+
+auto param = queryTest.getParameters();
+param.setValue("max_fragment_charge", 4);
+param.setValue("fragment_max_mz", 5000000); // That we definitively create all peptides
+queryTest.setParameters(param);
+StringList modifications_fixed_ = ListUtils::toStringList<std::string>(param.getValue("modifications_fixed"));
+StringList modifications_variable_ = ListUtils::toStringList<std::string>(param.getValue("modifications_variable"));
+ModifiedPeptideGenerator::MapToResidueType fixed_modifications = ModifiedPeptideGenerator::getModifications(modifications_fixed_);
+ModifiedPeptideGenerator::MapToResidueType variable_modifications = ModifiedPeptideGenerator::getModifications(modifications_variable_);
+
+//Create theoretical spectra for different charges
+TheoreticalSpectrumGenerator tsg;
+vector<AASequence> mod_peptides;
+PeakSpectrum b_y_ions;
+MSSpectrum spec_theo;
+Precursor prec_theo;
+
+
+
+vector<FragmentIndex::Peptide> peptides = queryTest.getPeptides();
+bool test = true;
+bool every_peak_found_its_counter_part = true;
+
+// Create different ms/ms spectra with different charges
+
+for(uint16_t charge = 1; charge <= 4; charge++)
+{
+  int peptide_idx = 0;
+// For each peptide that was created, we now generate a theoretical spectra for the given charge
+  //Each peptide should hit its own entry in the db. In this case the test returns true
+  for (auto& pep : peptides)
+  {
+    FragmentIndex::SpectrumMatchesTopN sms;
+    b_y_ions.clear(true);
+    mod_peptides.clear();
+    spec_theo.clear(true);
+
+    prec_theo.clearMetaInfo();
+    AASequence unmod_peptide = AASequence::fromString(entries[0].sequence.substr(pep.sequence_.first, pep.sequence_.second));
+    AASequence mod_peptide = AASequence(unmod_peptide); // copy the peptide
+    ModifiedPeptideGenerator::applyFixedModifications(fixed_modifications, mod_peptide);
+    ModifiedPeptideGenerator::applyVariableModifications(variable_modifications, mod_peptide, param.getValue("max_variable_mods_per_peptide"), mod_peptides);
+    mod_peptide = mod_peptides[pep.modification_idx_];
+    tsg.getSpectrum(b_y_ions, mod_peptide, charge, charge);
+    prec_theo.setMZ(mod_peptide.getMZ(charge));
+    spec_theo.setMSLevel(2);
```

Comment:
precursor charge not set here? not important?

Reply:
I wanted to test the scenario, in which the precursor charge is not known. In this case all different precursor charges are tested during the query (from min_precursor_charge to max_precursor_charge_).